apps/web/src/middlewares/ab-test-middleware.ts

```diff
@@ -0,0 +1,67 @@
+import { AB_TESTING_FEATURE_FLAG_MAP } from 'contexts/ABTestingContext/config'
+import { NextFetchEvent, NextRequest, NextResponse } from 'next/server'
+import { FeatureFlagInfo, MiddlewareFactory } from './types'
+
+// this function generates a deterministic result for a user for a given feature
+// it hashes a concatination of the users ip together with the features identifier and
+// probanility value. this allows us to ensure that a users probability result is different
+// for each feature gauranteeing a better distribution
+export const generateUserDeterministicValue = async (
+  userIp: string,
+  abTestingfeatureFlagInfo: FeatureFlagInfo[],
+): Promise<{ userWhitelistResults: { hasAccess: boolean; scaledValue: number }[] }> => {
```

Comment:
I think we only need to know whether the user has access to a specific feature or not

```suggestion
): Promise<{ [feature in EXPERIMENTAL_FEATURES]: boolean }> => {
```